File: leet/permutations-ii.ts
/**
 * https://leetcode.com/problems/permutations-ii/
 */
export function permuteUnique(nums: number[]): number[][] {
  let results: number[][] = [];

  let counter: Record<string, number> = {};
  for (let num of nums) {
    if (!(num in counter)) counter[num] = 0;
    counter[num]++;
  }

  function perm(elements: number[]) {
    if (elements.length === nums.length) {
      results.push([...elements]);
      return;
    }

    for (let num in counter) {
      if (counter[num] === 0) continue;

      elements.push(parseInt(num));
      counter[num]--;
      perm(elements);
      elements.pop();
      counter[num]++;
    }
  }

  perm([]);
  return results;
}

File: returned/python/review/permutations_ii.py
# https://leetcode.com/problems/permutations-ii/description/
from typing import List

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        result = []
        perm = []
        count = { n:0 for n in nums }
        for n in nums:
            count[n] += 1
        
        def dfs():
            if len(perm) == len(nums):
                result.append(perm[:])
                return
            
            for n in count:
                if count[n] > 0:
                    perm.append(n)
                    count[n] -= 1

                    dfs()

                    count[n] += 1
                    perm.pop()
        dfs()
        return result